### PR TITLE
[TRTLLM-10062][feat] Enable MTP for Nemotron  Super

### PR DESCRIPTION
**Pull request by sunnyqgg**
## Summary by CodeRabbit

* **New Features**
  * Speculative decoding support for Nemotron H with MTP layers
  * Triton-optimized causal convolution path for faster Mamba inference
  * Multi-token buffering and Eagle-tree-style attention retrieval for speculative flows

* **Improvements**
  * Preallocated output buffers to reduce memory copies and improve performance
  * Structured intermediate-state caching for verified speculative updates
  * Safer tile-size selection to avoid edge-case runtime errors

* **Tests**
  * Added MTP-enabled accuracy/config variants and a gated integration test (CI-skipped)

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by mikeiovine — COMMENTED** [review]
Great work @sunnyqgg. Shall we add an accuracy test to `tests/integration/defs/accuracy/test_llm_api_pytorch.py`? We can choose any small mamba hybrid for pre-merge and put the bigger ones in post-merge

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 1]
```diff
@@ -477,15 +486,274 @@ def __init__(
...
+            model=NemotronHModel(model_config),
+            model_config=model_config,
         )
+        self.model_nextn = 0
+        if model_config.spec_config is not None and model_config.spec_config.spec_dec_mode.is_mtp_one_model(
+        ):
+            model_nextn = model_config.spec_config.num_nextn_predict_layers
+            ckpt_nextn = self.config.num_nextn_predict_layers
+            self.num_hidden_layers = self.config.num_hidden_layers
+            assert ckpt_nextn > 0, "There is not MTP modules in the checkpoint."
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py]:
```suggestion
            assert ckpt_nextn > 0, "There are not MTP modules in the checkpoint."
```
Typo

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 2]
```diff
@@ -477,15 +486,274 @@ def __init__(
...
+                                    exclude_module.replace(
+                                        ckpt_prefix, model_prefix))
+                    self.model_config.quant_config.exclude_modules.extend(
+                        extend_exclude_modules)
+            self.model.layers.extend(self.draft_model.mtp_layers)
+            self.epilogue.extend(self.draft_model.mtp_layers)
+            self.epilogue.append(self.spec_worker)
 
-    def load_weights(self, weights: dict, weight_mapper: BaseWeightMapper):
+    def forward(
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py]: nit: if the method only forwards to the parent class, it can just be omitted entirely.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -1702,6 +1706,8 @@ def _executor_loop_overlap(self):
                     else:
                         previous_tensors_device = self.previous_batch and self.previous_batch.sample_state and self.previous_batch.sample_state.device
 
+                    if self.previous_batch is not None:
+                        self._update_mamba_cache_manager_for_previous_batch()
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: Does this need special handling in the pyexecutor? Where does rollback happen with overlap scheduling for normal linear KV cache?

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py [hunk 4]
```diff
@@ -55,6 +56,16 @@ def _split_mamba2_mixer_in_proj(w: torch.Tensor) -> torch.Tensor:
...
 
+            # MTP layers are stored as mtp.layers.0.xxx (sublayer 0, Attention) and mtp.layers.1.xxx (sublayer 1, MoE)
+            if "mtp.layers." in key:
+                import re
+                match = re.match(r'mtp\.layers\.(\d+)\.(.*)', key)
+                if match:
+                    sublayer_idx, rest = match.groups()
+                    key = f"model.layers.{config.num_hidden_layers}.layers.{sublayer_idx}.{rest}"
+                else:
+                    logger.warning(f"Failed to match MTP pattern for: {name}")
```
[the remark this answers is not in the document]
Comment (sunnyqgg) [reply tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: Yeah，it's better to error out, thanks, done.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py (hunk 2 above)
Comment (sunnyqgg) [reply tensorrt_llm/_torch/models/modeling_nemotron_h.py]: Done

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py [hunk 5]
```diff
@@ -239,38 +260,70 @@ def forward(
...
-            xbc_d = causal_conv1d_update(xbc_d,
-                                         conv_states,
-                                         self.conv1d.weight,
-                                         self.conv1d.bias,
-                                         activation="silu",
-                                         conv_state_indices=state_indices_d)
+            is_target_verify = attn_metadata.kv_cache_manager.is_speculative(
+            ) and spec_metadata is not None
+            if is_target_verify:
+                draft_token_num = spec_metadata.max_draft_len + 1
```
[the remark this answers is not in the document]
Comment (sunnyqgg) [reply tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: Done

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 6]
```diff
@@ -22,11 +24,54 @@
...
 from tensorrt_llm._torch.pyexecutor.scheduler import ScheduledRequests
 from tensorrt_llm.llmapi.llm_args import KvCacheConfig
+from tensorrt_llm.logger import logger
 from tensorrt_llm.mapping import Mapping
 
+if TYPE_CHECKING:
+    from tensorrt_llm._torch.attention_backend.interface import \
+        AttentionMetadata
+
+GB = 1 << 30
```
[the remark this answers is not in the document]
Comment (sunnyqgg) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: Sure,thanks,done

**Review by nv-guomingz — APPROVED** [review]

**Review by QiJune — APPROVED** [review]
LGTM

**Review by eopXD — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
